### PR TITLE
Fix LangChain autologging thread-safety issue

### PR DESCRIPTION
**Pull request by B-Step62**
### What changes are proposed in this pull request?

Resolve thread-safety issue in LangChain tracing (autologging). We found that the traces from Databricks agent evaluation are collapsed and the root cause was that `disable_autologging` is applied globally that does not work as agent eval runs model in a multiple worker thread.

Initial thought was that we don't need to disable tracing here, because MLflow has a guardrail for avoiding injecting callbacks twice. It checks if the tracing callback exists in the call args and only inject if not. Since LangChain propagates callbacks in args from parent to child components, this effectively avoid duplicate tracing for normal case.

However, I found one exception case, which is LangGraph. In LangGraph, users can define arbitrary python function as a graph node, which will be executed as a Runnable. The problematic case is where users invokes child Runnables within this custom code like this:

```
def my_custom_node(state):
     llm_response = ChatDatabricks.invoke(endpoint="...")
     # do something
     return state

graph.add_node("my_custom_node", my_custom_node) 
```

in this case, users don't propagate the callbacks to the `ChatDatabricks` call. Therefore, MLflow injects the callbacks when autologging is on. This itself might be OK, but what makes the problem worse is, LangChain actually propagates parent calbacks into `ChatDatabricks` here in different way than call args. It tracks parent callbacks in a context var and merge them into call args ([code](https://github.com/langchain-ai/langchain/blob/master/libs/core/langchain_core/runnables/config.py#L249)). This happens **after** our patching, so this will lead to the double tracing😞

Therefore, the only way I could find is disabling callback injection locally in that thread only. In other words, we disable artifact autologging and tracing in different scope:
* Artifact logging - disabled globally
* Trace logging - disabled locally in current thread.

I do not like the complexity of this solution .... but I could not come up with a better one and we need to resolve the issue asap.

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests

Multi-thread prediction generates correct traces:

```
with ThreadPoolExecutor(max_workers=3) as executor:
    futures = [executor.submit(predict, x=x) for x in range(10)]
    results = [f.result() for f in futures]
```
Screenshot 2024-11-05 at 21 53 11

Agent evaluation also produces correct traces:

```
with mlflow.start_run():
    mlflow.evaluate(model=langgraph_model, data=df, model_type="databricks-agent")
```
Screenshot 2024-11-05 at 21 48 45

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Fix thread-safety bug in LangChain autologging that potential generates collapsed traces when model runs in multiple threads.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [x] Yes (this PR will be cherry-picked and included in the next patch release)
- [ ] No (this PR will be included in the next minor release)

### CODE REVIEW

**Comment by WeichenXu123** [comment]
The overall idea makes sense. I will check details tomorrow.

File: mlflow/langchain/_langchain_autolog.py [hunk 1]
```diff
@@ -31,6 +32,46 @@
...
+# In normal cases, we do not need this flag because MLflow checks if the call args
+# already contains the MLflow callback before injecting it, and LangChain propagates
+# callbacks from parent to child, preventing double patching. However, an exception
+# is LangGraph, where users might define their custom python function and invoke
+# some LangChain classes manually without propagating the callbacks.
+IS_PATCHING_DISABLED_FOR_TRACING = ContextVar("is_trace_disabled", default=False)
+
+
+@contextlib.contextmanager
+def disable_patching():
```
Comment (WeichenXu123) [inline mlflow/langchain/_langchain_autolog.py:54]:
```suggestion
def _disable_patching():
```
this should be used internally

**Review by WeichenXu123 — APPROVED** [review]
LGTM!

alternative approach can be making tracing disablement flags thread-local, but patching ThreadPool API , when spawning new threads, copy thread-local tracing disablement flag value. but the approach is a bit tricky.

**Review by serena-ruan — APPROVED** [review]
LGTM! Could we drop `log_inputs_outputs` in autolog in a follow-up PR, it's been marked as deprecated for a long time :)

**Pull request merged**